syz-fuzzer/machine_info.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"runtime"
+	"strings"
+)
+
+func CollectMachineInfo() ([]byte, error) {
+	if runtime.GOOS != "linux" {
+		return []byte{}, nil
+	}
+
+	type machineInfoFunc struct {
+		name string
+		fn   func() ([]byte, error)
+	}
+
+	allMachineInfo := []machineInfoFunc{
+		{"CPU Info", readCPUInfo},
+		{"KVM", readKVMInfo},
+	}
+
+	var buffer bytes.Buffer
+
+	for _, pair := range allMachineInfo {
+		buffer.WriteString(fmt.Sprintf("[%s]\n", pair.name))
+		data, err := pair.fn()
+		if err != nil {
+			if os.IsNotExist(err) {
+				buffer.WriteString(err.Error() + "\n")
+			} else {
+				return nil, err
+			}
+		} else {
+			buffer.Write(data)
+		}
+	}
+
+	return buffer.Bytes(), nil
+}
+
+func readCPUInfo() (result []byte, err error) {
+	result, err = ioutil.ReadFile("/proc/cpuinfo")
```

Comment:
`return ioutil.ReadFile("/proc/cpuinfo")`

Reply:
But I am not sure we need all of it. Think of machine with 256 cores. Info for each of them will be almost identical, except for NUMA info, which is better sourced from some other location (or we select just the relevant lines).
E.g. on my machine it's 120K and 2000+ lines. Think of scrolling to the bottom of it each time.

Reply:
I tried to parse `/proc/cpuinfo`. The format differs from different types of CPUs, but most of them seems to have the format:
```
[general info]
[cpu 0 info]
[cpu 1 info]
...
```
All the keys inside the section of each cpu is the same, so I group the lines by its key. If they all have the same value then I will only print one. If they have different values I will print them in one line seperated by comma.
Let's see if this work well enough.
Not sure how to test this with different cpus.